src/main/scala/org/economicsl/auctions/sandbox.sc

```diff
@@ -53,4 +53,24 @@ val orderBook5 = orderBook4 + order8
 
 // take a look at paired orders
 val (pairedOrders, _) = orderBook5.takeWhileMatched
-pairedOrders.toList
\ No newline at end of file
+pairedOrders.toList
+
+
+// Why can this not be made covariant in T? Can divisible orders be constructed using an auxiliary constructor on LimitAskOrder?
+case class DivisibleLimitAskOrder[T <: Tradable](issuer: UUID, limit: Price, quantity: Quantity, tradable: T) extends
+  LimitAskOrder[T] with Divisible[T, DivisibleLimitAskOrder[T]] {
+
+  def split(residual: Quantity): (DivisibleLimitAskOrder[T], DivisibleLimitAskOrder[T]) = {
+    require(residual.value < this.quantity.value)  // can this be check be lifted into the type system?
```

Comment:
@bherd-rb I am pretty sure that this sanity check can not be performed by the compiler because the value passed to the method is inherently run-time specific.  What we could do is, rather than use an assert/require statement to throw an exception, we could change the type signature to be

```scala
def split(residual: Quantity): Try[(DivisibleLimitAskOrder[T], DivisibleLimitAskOrder[T])]
```

where the [`Try`](http://www.scala-lang.org/api/current/scala/util/Try.html) type would return `Success(DivisibleLimitAskOrder[T], DivisibleLimitAskOrder[T])` in the event that the requirement is satisfied and `Failure(SomeCustomException)` in the event that the requirement has not been satisfied.

For research simulations, I think using the `require` statement is the way to go because if the caller of the `split` method is violating the requirement it indicates a bug in the calling code.  In an IoT production environment I am less sure that the `require` statement is the right approach.  I think it depends on where the calling code resides.  At present the `split` method would only ever be used by the `FourHeapOrderBook` (or other `OrderBook` classes) to split and order during the clearing process.

Actually, perhaps this indicates that the `split` method should be a private method of the `FourHeapOrderBook` class?